File: src/darkyenus/resourcepacker/tasks/PackTask.scala
package darkyenus.resourcepacker.tasks

import java.awt.Color
import java.io.FileReader
import javax.imageio.ImageIO

import com.badlogic.gdx.tools.texturepacker.TexturePacker
import com.badlogic.gdx.utils.{Json, JsonReader}
import com.esotericsoftware.minlog.Log
import com.google.common.io.Files
import darkyenus.resourcepacker.util.ImageUtil
import darkyenus.resourcepacker.{ResourceDirectory, ResourceFile, Task}

/**
 * Packs all images in .pack. flagged directory using libGDX's texture packer and then flattens it.
 * Can also preblend all packed resources if supplied with .#RRGGBB. flag, see [[PreBlendTask]].
 *
 * If the directory contains pack.json file (name can contain flags),
 * it will be used in packing, as with default packing procedures.
 *
 * @author Darkyen
 */
object PackTask extends Task {

  /**
   * Matches: #RRGGBB
   * Where RR (GG and BB) are hexadecimal digits.
   * Example:
   * #FF0056
   * to capture groups RR GG BB
   */
  val PreBlendRegex = "#([0-9A-Fa-f][0-9A-Fa-f])([0-9A-Fa-f][0-9A-Fa-f])([0-9A-Fa-f][0-9A-Fa-f])".r

  private val json = new Json()
  private val jsonReader = new JsonReader

  override def operate(packDirectory: ResourceDirectory): Boolean = {
    if (packDirectory.flags.contains("pack")) {
      val settings = new TexturePacker.Settings()
      settings.limitMemory = false
      settings.useIndexes = false
      settings.filterMag = com.badlogic.gdx.graphics.Texture.TextureFilter.Linear
      settings.filterMin = settings.filterMag
      settings.pot = true //Seems that it is still better for performance and whatnot
      settings.maxWidth = 2048
      settings.maxHeight = settings.maxWidth
      settings.duplicatePadding = true
      settings.bleed = true
      settings.stripWhitespaceX = true
      settings.stripWhitespaceY = true
      settings.alphaThreshold = 0
      packDirectory.files.find(file => file.name == "pack" && file.extension == "json") match {
        case Some(packFile) =>
          json.readFields(settings, jsonReader.parse(new FileReader(packFile.file)))
          packDirectory.removeChild(packFile)
          Log.info(Name, "Json packer settings loaded. " + packFile.file.getAbsolutePath)
        case None =>
      }
      val packer = new TexturePacker(settings)
      for (image <- packDirectory.files if image.isImage) {
        val bufferedImage = ImageIO.read(image.file)
        if (bufferedImage == null) {
          Log.error(Name, "Image could not be loaded. " + image)
        } else {
          packer.addImage(bufferedImage, image.name + (if (image.flags.contains("9")) ".9" else ""))
          Log.info(Name, s"Image added to pack. " + image)
        }
        packDirectory.removeChild(image)
      }

      val atlasName = packDirectory.name
      val outputFolder = newFolder()
      packer.pack(outputFolder, atlasName)

      val preBlendColor = packDirectory.flags collectFirst {
        case PreBlendRegex(r, g, b) =>
          Log.info(Name, "Blending color for atlas set. " + atlasName)
          new Color(Integer.parseInt(r, 16), Integer.parseInt(g, 16), Integer.parseInt(b, 16))
      }

      for (outputJavaFile <- outputFolder.listFiles()) {
        if (Files.getFileExtension(outputJavaFile.getName).equalsIgnoreCase("png")) {
          for (color <- preBlendColor) {
            ImageUtil.preBlendImage(outputJavaFile, color)
          }
        }
        packDirectory.parent.addChild(new ResourceFile(outputJavaFile, packDirectory.parent))
      }
      FlattenTask.flatten(packDirectory)
      Log.info(Name, "Image atlas packed and directory has been flattened. " + atlasName + " " + packDirectory)
      true
    } else false
  }
}

File: src/main/java/com/darkyen/resourcepacker/tasks/PackTask.kt
package com.darkyen.resourcepacker.tasks

import com.badlogic.gdx.graphics.Texture.TextureFilter.Linear
import com.badlogic.gdx.utils.Json
import com.badlogic.gdx.utils.JsonReader
import com.darkyen.resourcepacker.Resource
import com.darkyen.resourcepacker.Resource.ResourceDirectory
import com.darkyen.resourcepacker.Task
import com.darkyen.resourcepacker.image.createImage
import com.darkyen.resourcepacker.isImage
import com.darkyen.resourcepacker.util.texturepacker.MultiScaleTexturePacker
import com.darkyen.resourcepacker.util.texturepacker.MultiScaleTexturePacker.Settings
import com.esotericsoftware.minlog.Log
import java.io.FileReader

/**
 * Packs all images in .pack. flagged directory using libGDX's texture packer and then flattens it.
 *
 * If the directory contains pack.json file (name can contain flags),
 * it will be used in packing, as with default packing procedures.
 *
 * Can generate multiple atlas image files with different, Apple-like densities, that is, with @2x scheme.
 * Files with @Nx (where N is scale level) are assigned to that level.
 * Missing density levels are created from existing ones, either by up/downscaling or by rasterization with different dimensions.
 * Automatically rasterizes .svg files, flags from RasterizeTask can be used for specifying scale.
 *
 * Generates only one atlas file and (potentially) multiple image files.
 * Therefore, all images are packed to the same locations, only all coordinates are multiplied by N.
 * This allows for simple runtime density switching.
 *
 * Scale level 1 is specified automatically, others (usually 2) can be added with @Nx flag (@2x in case of level 2).
 * Note that if the level is not a power of two (for example 3), `pot` setting must be set to false (default is true),
 * otherwise the level will be ignored and warning will be emitted.
 *
 * @author Darkyen
 */
object PackTask : Task() {

    private val ScalesRegex = Regex("@([1-9]+[0-9]*)x")

    private val ScaledNameRegex = Regex("(.+)@([1-9]+[0-9]*)x?")

    private val json = Json()
    private val jsonReader = JsonReader()

    override fun operate(directory: ResourceDirectory): Boolean {
        if (!directory.flags.contains("pack")) {
            return false
        }

        val settings = Settings()
        settings.limitMemory = false
        settings.useIndexes = false
        settings.filterMag = Linear
        settings.filterMin = settings.filterMag
        settings.pot = true //Seems that it is still better for performance and whatnot
        settings.maxWidth = 2048
        settings.maxHeight = settings.maxWidth
        settings.duplicatePadding = true
        settings.bleed = true
        settings.stripWhitespaceX = true
        settings.stripWhitespaceY = true
        settings.alphaThreshold = 0
        settings.ignoreBlankImages = false
        //settings.debug = true
        for (packFile in directory.files) {
            if (packFile.name == "pack" && packFile.extension == "json") {
                json.readFields(settings, jsonReader.parse(FileReader(packFile.file)))
                directory.removeChild(packFile)
                if (Log.DEBUG) Log.debug(Name, "Json packer settings loaded. " + packFile.file.canonicalPath)
                break
            }
        }

        if (settings.duplicatePadding && (settings.paddingX < 2 && settings.paddingY < 2)) {
            Log.warn(Name, "duplicatePadding is true, but padding is less than 2 so it won't have any effect")
        } else if (settings.duplicatePadding && (settings.paddingX < 2 || settings.paddingY < 2)) {
            if (Log.DEBUG) Log.debug(Name,
                    "duplicatePadding is true, but padding of one dimension is less than 2 so it won't have any effect " +
                            "(paddingX = " + settings.paddingX + ", paddingY = " + settings.paddingY + ")")
        }

        val scales = com.badlogic.gdx.utils.IntArray()
        scales.add(1)
        for (flag in directory.flags) {
            val scaleMatch = ScalesRegex.matchEntire(flag) ?: continue
            val scale = scaleMatch.groupValues[1].toInt()
            if (!scales.contains(scale)) {
                scales.add(scale)
            }
        }
        scales.sort()
        settings.scales = scales.toArray()

        val packer = MultiScaleTexturePacker(settings)
        directory.forEachFile(filter = { it.isImage() }) { image ->
            var name = image.name
            var scale = 1

            val scaledNameMatch = ScaledNameRegex.matchEntire(name)
            if (scaledNameMatch != null) {
                name = scaledNameMatch.groupValues[1]
                scale = scaledNameMatch.groupValues[2].toInt()
            }

            packer.addImage(name, -1, scale, image.createImage())

            if (Log.DEBUG) Log.debug(Name, "Image added to pack. $image")
            directory.removeChild(image)
        }

        val atlasName = directory.name
        val outputFolder = newFolder()
        packer.pack(outputFolder, atlasName)


        for (outputJavaFile in outputFolder.listFiles() ?: emptyArray()) {
            directory.parent.addChild(Resource.ResourceFile(outputJavaFile, directory.parent))
        }
        FlattenTask.flatten(directory)
        Log.info(Name, "Image atlas packed and directory has been flattened. $atlasName $directory")
        return true
    }
}
